File: ethutil/natspec/natspec.go
package natspec

import (
	"encoding/json"
	// "fmt"
	"github.com/ethereum/go-ethereum/javascript"
)

type ptype int

/*
NatSpec processes contract metadata
initialised from JSON-format contract definitions

 def: contract ABI definition
    - https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI#json
 doc: contract user documentation
    - https://github.com/ethereum/wiki/wiki/Ethereum-Natural-Specification-Format#user-documentation
 def: contract ABI definition
    - https://github.com/ethereum/wiki/wiki/Ethereum-Natural-Specification-Format#developer-documentation
*/
type NatSpec struct {
	def     [](map[string]interface{})
	doc     map[string]interface{}
	dev     map[string]interface{}
	methods map[string]*meth
	jsre    *javascript.JSRE
}

// This can be cached and called multiple times when the method is called with different parameters.
func New(defJson, docJson []byte, jsreF func() (*javascript.JSRE, error)) (self *NatSpec, err error) {
	def := []map[string](interface{}){}
	if err = json.Unmarshal(defJson, &def); err != nil {
		return
	}

	doc := make(map[string]interface{})
	if err = json.Unmarshal(docJson, &doc); err != nil {
		return
	}
	self = &NatSpec{def, doc, nil, make(map[string]*meth), nil}
	if self.jsre, err = jsreF(); err != nil {
		return
	}
	return
}

type meth struct {
	args  []string
	types []ptype
	// signature string
	// key       string
	notice string
}

// var m = &method{}

func (self *NatSpec) Notice(method string, params ...interface{}) (notice string, err error) {
	m := self.methods[method]
	if m == nil {
		var info, param map[string]interface{}
		for _, info = range self.def {
			name, ok := info["name"].(string)
			if !ok {
				return
			}
			if name == method {
				break
			}
		}
		var args []string
		var types []ptype
		inputs, ok := info["inputs"].([]map[string](interface{}))
		if !ok {
			return
		}
		var name string
		var typ ptype
		for _, param = range inputs {
			name, ok = param["name"].(string)
			args = append(args, name)
			typ, ok = param["type"].(ptype)
			types = append(types, typ)
		}
		m = &meth{args, types /* signature, key, */, notice}
		self.methods[method] = m
	}
	// key := crypto.Sha3(signature)
	key := ""
	methods, ok := self.doc["methods"].(map[string](map[string](interface{})))
	if !ok {
		return
	}
	meth := methods[key]
	// meth, ok := methods[key].(map[string](interface{}))
	if !ok {
		return
	}
	expression, ok := meth["notice"].(string)
	notice = expression
	// notice = "" // evaluate(expression)
	return
}

var evalScript = "// match everything in `` quotes" +
	"var pattern = /\\\\`(?:\\\\.|[^`\\\\])*\\\\`/gim" +
	`var match;
var lastIndex = 0;
while ((match = pattern.exec(expression)) !== null) {
  var startIndex = pattern.lastIndex - match[0].length;
  var toEval = match[0].slice(1, match[0].length - 1);
  evaluatedExpression += expression.slice(lastIndex, startIndex);
  evaluatedExpression += eval(toEval).toString();
  lastIndex = pattern.lastIndex;
}
evaluatedExpression += expression.slice(lastIndex);
return evaluatedExpression;
`

	// @title: This is a title that should describe the contract and go above the contract definition
	// @author: The name of the author of the contract. Should also go above the contract definition.
	// @notice: Represents user documentation. This is the text that will appear to the user to notify him of what the function he is about to execute is doing
	// @dev: Represents developer documentation. This is documentation that would only be visible to the developer.
	// @param: Documents a parameter just like in doxygen. Has to be followed by the parameter name.
	// @return: Documents the return type of a contract's function.

File: ethutil/natspec/natspec.js
/*
    This file is part of natspec.js.

    natspec.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    natspec.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with natspec.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/** @file natspec.js
 * @authors:
 *   Marek Kotewicz <marek@ethdev.com>
 * @date 2015
 */

//var abi = require('./node_modules/ethereum.js/lib/abi.js'); 
//var abi = require('./lib/abi.js'); 

/**
 * This object should be used to evaluate natspec expression
 * It has one method evaluateExpression which shoul be used
 */
var natspec = (function () {
    /// Helper method
    /// Should be called to copy values from object to global context
    var copyToContext = function (obj, context) {
        var keys = Object.keys(obj);
        keys.forEach(function (key) {
            context[key] = obj[key];
        });
    }

    /// this function will not be used in 'production' natspec evaluation
    /// it's only used to enable tests in node environment
    /// it copies all functions from current context to nodejs global context
    var copyToNodeGlobal = function (obj) {
        if (typeof global === 'undefined') {
            return;
        }
        copyToContext(obj, global);
    };

    /// Helper method
    /// Should be called to get method with given name from the abi
    /// @param contract's abi
    /// @param name of the method that we are looking for
    var getMethodWithName = function(abi, name) {
        return abi.filter(function (method) {
            return method.name === name;
        })[0];
    };

    /// Function called to get all contract's storage values
    /// @returns hashmap with contract properties which are used
    /// TODO: check if this function will be used
    var getContractProperties = function (address, abi) {
        return {};
    };

    /// Function called to get all contract method input variables
    /// @returns hashmap with all contract's method input variables
    var getMethodInputParams = function (method, transaction) {
        // do it with output formatter (cause we have to decode)
        var params = abi.formatOutput(method.inputs, '0x' + transaction.params[0].data.slice(10)); 

        return method.inputs.reduce(function (acc, current, index) {
            acc[current.name] = params[index];
            return acc;
        }, {});
    };
    
    /// Should be called to evaluate single expression
    /// Is internally using javascript's 'eval' method
    /// @param expression which should be evaluated
    /// @param [call] object containing contract abi, transaction, called method
    /// TODO: separate evaluation from getting input params, so as not to spoil 'evaluateExpression' function
    var evaluateExpression = function (expression, call) {
        var self = this;
        
        if (!!call) {
            try {
                var method = getMethodWithName(call.abi, call.method);
                var params = getMethodInputParams(method, call.transaction); 
                copyToContext(params, self);
            }
            catch (err) {
                return err;//"Natspec evaluation failed, wrong input params";
            }
        }

        // used only for tests
        copyToNodeGlobal(self);

        var evaluatedExpression = "";

        // match everything in `` quotes
        var pattern = /\`(?:\\.|[^`\\])*\`/gim
        var match;
        var lastIndex = 0;
        while ((match = pattern.exec(expression)) !== null) {
            var startIndex = pattern.lastIndex - match[0].length;

            var toEval = match[0].slice(1, match[0].length - 1);

            evaluatedExpression += expression.slice(lastIndex, startIndex);

            var evaluatedPart;
            try {
                evaluatedPart = eval(toEval).toString(); 
            }
            catch (err) {
                evaluatedPart = 'undefined'; 
            }

            evaluatedExpression += evaluatedPart;
            lastIndex = pattern.lastIndex;
        }

        evaluatedExpression += expression.slice(lastIndex);
        
        return evaluatedExpression;
    };

    return {
        evaluateExpression: evaluateExpression
    };

})();

//module.exports = natspec; 

